controller/backup_volume_controller.go

```diff
@@ -0,0 +1,375 @@
+	if err == nil {
+		bvc.queue.Forget(key)
+		return
+	}
+
+	if bvc.queue.NumRequeues(key) < maxRetries {
+		bvc.logger.WithError(err).Warnf("Error syncing Longhorn backup volume %v", key)
+		bvc.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	bvc.logger.WithError(err).Warnf("Dropping Longhorn backup volume %v out of the queue", key)
+	bvc.queue.Forget(key)
+}
+
+func (bvc *BackupVolumeController) DeleteBackups(volumeName string, fileCleanupRequired bool) error {
+	backups, err := bvc.ds.ListBackup(volumeName)
+	if err != nil {
+		return err
+	}
+
+	var errs []string
+	if fileCleanupRequired {
+		for _, backup := range backups {
+			backup.Spec.FileCleanupRequired = true
+			if _, err = bvc.ds.UpdateBackup(backup); err != nil && !datastore.ErrorIsConflict(err) {
+				errs = append(errs, err.Error())
+				continue
+			}
+		}
+		if len(errs) > 0 {
+			return errors.New(strings.Join(errs, ","))
+		}
+	}
+
+	for backupName := range backups {
+		err = bvc.ds.DeleteBackup(backupName)
+		if err != nil {
+			errs = append(errs, err.Error())
+			continue
+		}
+	}
+	if len(errs) > 0 {
+		return errors.New(strings.Join(errs, ","))
+	}
+	return nil
+}
+
+func (bvc *BackupVolumeController) reconcile(log logrus.FieldLogger, backupVolumeName string) (err error) {
+	backupTarget, err := bvc.ds.GetBackupTarget(defaultBackupTargetName)
+	if err != nil {
+		if !datastore.ErrorIsNotFound(err) {
+			return err
+		}
+		log.Warnf("Backup target %s be deleted", defaultBackupTargetName)
+		return nil
+	}
+
+	// Check the responsible node
+	if backupTarget.Spec.ResponsibleNodeID != bvc.controllerID {
+		return nil
+	}
+
+	log = log.WithFields(logrus.Fields{
+		"backupVolume": backupVolumeName,
+	})
+
+	// Get BackupVolume CR
+	backupVolume, err := bvc.ds.GetBackupVolume(backupVolumeName)
+	if err != nil {
+		if !datastore.ErrorIsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	// Examine DeletionTimestamp to determine if object is under deletion
+	if backupVolume.DeletionTimestamp != nil {
+		// Delete related Backup CRs with the given backup volume name
+		if err := bvc.DeleteBackups(backupVolumeName, backupVolume.Spec.FileCleanupRequired); err != nil {
+			log.WithError(err).Error("Error deleting backups")
+			return err
+		}
+
+		// Delete the backup volume from the remote backup target
+		if backupVolume.Spec.FileCleanupRequired {
+			backupTargetClient, err := manager.GenerateBackupTarget(bvc.ds)
+			if err != nil {
+				log.WithError(err).Error("Error generate backup target client")
+				// Ignore error to prevent enqueue
+				return nil
+			}
+
+			if err := backupTargetClient.DeleteBackupVolume(backupVolumeName); err != nil {
+				log.WithError(err).Error("Error deleting remote backup volume")
+				return err
+			}
+		}
+
+		// Clean up Volume CR status
+		if err := manager.SyncVolumeLastBackupWithBackupVolume(backupVolumeName, nil, bvc.ds.GetVolume, bvc.ds.UpdateVolumeStatus); err != nil {
```

Comment:
I think it's VolumeController's responsibility to sync volume CRs with backupVolume CRs now.